patches/server/0739-Add-CompostItemEvent.patch

```diff
@@ -0,0 +1,103 @@
+--- a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
+@@ -219,7 +219,12 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+ 
+         if (i < 8 && ComposterBlock.COMPOSTABLES.containsKey(itemstack.getItem())) {
+             if (i < 7 && !world.isClientSide) {
+-                BlockState iblockdata1 = ComposterBlock.addItem(state, (LevelAccessor) world, pos, itemstack);
++                // Paper start
++                BlockState iblockdata1 = ComposterBlock.addItem(state, (LevelAccessor) world, pos, itemstack, player);
++                if (iblockdata1 == null) {
++                    return InteractionResult.PASS;
++                }
++                // Paper start
+ 
+                 world.levelEvent(1500, pos, state != iblockdata1 ? 1 : 0);
+                 player.awardStat(Stats.ITEM_USED.get(itemstack.getItem()));
+@@ -243,11 +248,16 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+         if (i < 7 && ComposterBlock.COMPOSTABLES.containsKey(itemstack.getItem())) {
+             // CraftBukkit start
+             double rand = worldserver.getRandom().nextDouble();
+-            BlockState iblockdata1 = ComposterBlock.addItem(iblockdata, DummyGeneratorAccess.INSTANCE, blockposition, itemstack, rand);
++            BlockState iblockdata1 = ComposterBlock.addItem(iblockdata, DummyGeneratorAccess.INSTANCE, blockposition, itemstack, rand, entity); // Paper
+             if (iblockdata == iblockdata1 || org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, iblockdata1).isCancelled()) {
+                 return iblockdata;
+             }
+-            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand);
++            // Paper start
++            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand, entity);
++            if (iblockdata1 == null) {
++                return iblockdata;
++            }
++            // Paper end
+             // CraftBukkit end
+ 
+             itemstack.shrink(1);
+@@ -291,16 +301,44 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+     }
+ 
+     static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item) {
+-        // CraftBukkit start
+-        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
++        // Paper start
++        return ComposterBlock.addItem(state, world, pos, item, null);
++    }
++    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
++        // Paper end
++        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
```

Comment:
```suggestion
         // CraftBukkit start
         return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
     }
+    // Paper start
+    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
+        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
+    }
+    // Paper end
```

This block of diff can be simplified. (Reminder not to use github's commit from suggestion feature, this is just to show what it should sorta look like)